tools/PI/DevHome.PI/Models/WERHelper.cs

```diff
@@ -0,0 +1,541 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Diagnostics.Eventing.Reader;
+using System.IO;
+using System.Threading;
+using DevHome.Common.Helpers;
+using Microsoft.Win32;
+using Serilog;
+
+namespace DevHome.PI.Models;
+
+/*
+ * This class is responsible for monitoring the system for WER reports. These reports are generated when an application crashes.
+ * Additionally it can be used to enable/disable local WER collection for a specific app. To learn more about local WER collection,
+ * check out https://learn.microsoft.com/windows/win32/wer/collecting-user-mode-dumps
+ *
+ * We learn about WER events from either events in the Application event log or from crash dump files on disk.
+ */
+
+internal sealed class WERHelper : IDisposable
+{
+    private const string WERSubmissionQuery = "(*[System[Provider[@Name=\"Application Error\"]]] and *[System[EventID=1000]])";
+    private const string WERReceiveQuery = "(*[System[Provider[@Name=\"Application Error\"]]] and *[System[EventID=1001]])";
+    private const string DefaultDumpPath = "%LOCALAPPDATA%\\CrashDumps";
+    private const string LocalWERRegistryKey = "SOFTWARE\\Microsoft\\Windows\\Windows Error Reporting\\LocalDumps";
+
+    private static readonly ILogger _log = Log.ForContext("SourceContext", nameof(WERHelper));
+
+    private readonly Microsoft.UI.Dispatching.DispatcherQueue _dispatcher;
+    private readonly EventLogWatcher _eventLogWatcher;
+    private readonly List<FileSystemWatcher> _filesystemWatchers = [];
+    private readonly ObservableCollection<WERReport> _werReports = [];
+
+    private List<string> _werLocations = [];
```

Comment:
Excellent question. Of the lists shown here, _werReports is the one most interesting and subject to problems. I thought I had it taken care of, but I looked at it again and had a problem.

Since I'm databinding to this list, all of the clients were accessing it via the UI thread, and thus WERHelper was only adding/modifying the list on the dispatcher thread as well. However, there were a couple of functions that were enumerating the list in WERHelper on a ThreadPool thread, and that could run into issues. I originally was doing some locking to protect the list in the early implementation, but that locking was pretty useless when there were other clients accessing the list without acquiring locks. I've updated the code to only access that list from the Dispatcher thread, so it should be safe now.